File: src/omnia/data/structure/tuple/Triplet.java
package omnia.data.structure.tuple;

import java.util.function.Function;

public interface Triplet<T> extends Triple<T, T, T>, Tuples.AtMostTriplet<T> {

  static <T> Triplet<T> of(T first, T second, T third) {
    return new ImmutableTriplet<>(first, second, third);
  }

  @Override
  <R> Triplet<R> map(Function<? super T, ? extends R> mapper);

  @Override
  Couplet<T> dropFirst();

  @Override
  Couplet<T> dropSecond();

  @Override
  Couplet<T> dropThird();

  @Override
  Quadruplet<T> concat(T object);

  @Override
  Quintuplet<T> concat(Couple<T, T> other);

  @Override
  Sextuplet<T> concat(Triple<T, T, T> other);

  @Override
  Septuplet<T> concat(Quadruple<T, T, T, T> other);

  @Override
  Octuplet<T> concat(Quintuple<T, T, T, T, T> other);

  @Override
  Nonuplet<T> concat(Sextuple<T, T, T, T, T, T> other);

  @Override
  Decuplet<T> concat(Septuple<T, T, T, T, T, T, T> other);
}

File: src/omnia/data/structure/tuple/Triplet.kt
package omnia.data.structure.tuple

import omnia.data.structure.tuple.Tuples.AtMostTriplet
import java.util.function.Function

interface Triplet<T> : Triple<T, T, T>, AtMostTriplet<T> {
    override fun <R> map(mapper: Function<in T, out R>): Triplet<R>
    override fun dropFirst(): Couplet<T>
    override fun dropSecond(): Couplet<T>
    override fun dropThird(): Couplet<T>
    override fun concat(`object`: T): Quadruplet<T>
    override fun concat(other: Couple<T, T>): Quintuplet<T>
    override fun concat(other: Triple<T, T, T>): Sextuplet<T>
    override fun concat(other: Quadruple<T, T, T, T>): Septuplet<T>
    override fun concat(other: Quintuple<T, T, T, T, T>): Octuplet<T>
    override fun concat(other: Sextuple<T, T, T, T, T, T>): Nonuplet<T>
    override fun concat(other: Septuple<T, T, T, T, T, T, T>): Decuplet<T>

    companion object {
        fun <T> of(first: T, second: T, third: T): Triplet<T> {
            return ImmutableTriplet(first, second, third)
        }
    }
}